core/arch/riscv64/mangle.c

```diff
@@ -612,37 +612,19 @@ instr_t *
 mangle_rel_addr(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
                 instr_t *next_instr)
 {
-    opnd_t dst = instr_get_dst(instr, 0);
     app_pc tgt;
     ASSERT(instr_get_opcode(instr) == OP_auipc);
     ASSERT(instr_has_rel_addr_reference(instr));
     instr_get_rel_data_or_instr_target(instr, &tgt);
-    ASSERT(opnd_is_reg(dst));
-    ASSERT(opnd_is_rel_addr(instr_get_src(instr, 0)));
-    bool uses_reg_stolen = instr_uses_reg(instr, dr_reg_stolen),
-         uses_reg_tp = instr_uses_reg(instr, DR_REG_TP);
 
-    if (uses_reg_stolen || uses_reg_tp) {
-        dst = opnd_create_reg(DR_REG_A0);
-        PRE(ilist, next_instr,
-            instr_create_save_to_tls(dcontext, DR_REG_A0, TLS_REG0_SLOT));
-    }
+    mangle_special_registers(dcontext, ilist, instr, next_instr);
```

Comment:
Best to get the return value and pass it on in case something inside here removes next_instr.

Reply:
next_instr isn't used anymore in the function, so I think it's unneeded. And we cannot return a non-NULL pointer in `mangle_rel_addr()`, since we delete the original `instr` and must inform the caller with a NULL return value[1]. Maybe we should add a note of not removing `next_instr` to `mangle_special_registers()`?

[1]: https://github.com/DynamoRIO/dynamorio/blob/0ab4d54744519e01854784612a83ddc57a16b0ad/core/arch/mangle_shared.c#L1953-L1965

Reply:
This seems dangerous vs future changes: IMHO either mangle_special_registers() should not return a new next_instr then, so it cannot remove it, or else this function needs to return next_instr so it can pass on the return value from mangle_special_registers().

Reply:
I switch to use mangle_stolen_reg_and_tp_reg(), which doesn't return a new next_instr